### ISSUE TITLE
Allow black box ops (with custom_gradient) in @tf.function

### ISSUE THREAD

**Issue by ziofil** [tensorflow/tensorflow#41897]
**System information**
- TensorFlow version (you are using): 2.3
- Are you willing to contribute it (Yes/No): No (don't know how)

**Describe the feature and the current behavior/state.**
When decorating a function with `@tf.function` it would be great to be able to include untraceable ops (e.g., because they contain compiled numba code), as long as they are decorated with  `@tf.custom_gradient` and somehow their signature and dtype is well specified.

Unless I'm missing something major, once a function is decorated with `@custom_gradient`, it shouldn't be necessary for AutoGraph to trace it. It should be "pluggable" in the computational graph as a block.

**Will this change the current api? How?**
Possibly (perhaps optional flags in `tf.function`).

**Who will benefit with this feature?**
Users who want to use custom/compiled code compatibly with `tf.function`/autograph (which sometimes happens by default, e.g. when using `Model.fit()` in keras).

**Comment by mdanatg** [comment]
For Python interop, [py_function](https://www.tensorflow.org/api_docs/python/tf/py_function) is effectively a black box. Beware however that the overhead of calling back into the Python runtime is not negligible. `py_function` is also not portable and is not saved by saved_model. `py_function` itself doesn't allow a custom gradients, but you can always write a wrapper function that has a @custom_gradient.

For C++ interop, there is the alternative of registering a custom op kernel.

Feel free to reopen the issue if the options above don't work for you.
